sns_ik_lib/include/sns_ik/sns_ik_base.hpp

```diff
@@ -0,0 +1,248 @@
+   */
+  const Eigen::ArrayXd& getUpp() const { return qUpp_; };
+
+  /*
+   * @return: number of joints
+   */
+  int nJnt() const { return nJnt_; }
+
+protected:
+
+  /*
+   * The code of the SNS-IK solver relies on a linear solver. If the linear system is infeasible,
+   * then the solver will return the minimum-norm solution with a non-zero (positive) residual.
+   * This corresponds to a situation where there is no feasible solution to the task objective, for
+   * example, a one-link robot arm where the task velocity is along the axis of the joint. This
+   * threshold is used to determine if the task is infeasible.
+   */
+  static const double LIN_SOLVE_RESIDUAL_TOL;
+
+  /*
+   * If all goes well, then the solver should always terminate the main loop. If something crazy
+   * happens, then we want to prevent an infinite loop. The maximum iteration count is given by
+   * this factor multiplied by the number of joints in the task jacobian.
+   */
+  static const int MAXIMUM_SOLVER_ITERATION_FACTOR;
+
+  // Short names for useful constants
+  static const double POS_INF;
+  static const double NEG_INF;
+
+  // Any number smaller than this is considered to be zero for numerical purposes
+  static const double MINIMUM_FINITE_SCALE_FACTOR;
+
+  // Any number larger than this is considered to be infinite for numerical purposes
+  static const double MAXIMUM_FINITE_SCALE_FACTOR;
+
+  // Tolerance for checks on the velocity/acceleration limits
+  static const double BOUND_TOLERANCE;
+
+  // Nice formatting option from Eigen
+  static const Eigen::IOFormat EigArrFmt;
+
+  /*
+   * protected constructor: require factory method to create an object.
+   */
+  SnsIkBase(int nJnt) : nJnt_(nJnt), qLow_(nJnt), qUpp_(nJnt) {};
+
+  /*
+   * Check that qLow_ <= q <= qUpp_
+   * @param qUpp: joint acceleration to test
+   * @return: true iff qLow <= q <= qUpp
+   *          if q.empty() return false
+   */
+  bool checkBnd(const Eigen::VectorXd& q);
+
+  /*
+   * This method sets and solves the decomposition of the matrix that is used by the linear solver.
+   * In general, this will be the J*W matrix, which describes the jacobian of the active joints.
+   * @param JW: matrix to set in the linear solver.
+   * @return: Success if the decomposition was successful
+   */
+  SnsIkBase::ExitCode setLinearSolver(const Eigen::MatrixXd& JW);
+
+  /*
+   * Solve a specific linear system and compute the residual error. Linear system:
+   * JW * q = rhs
+   * @param rhs: "right hand side" of the linear system.
+   * @param[out] q: solution to the linear system
+   * @param[out] resErr: residual error in the linear system
+   * @return: Success if the solve was successful
+   */
+  SnsIkBase::ExitCode solveLinearSystem(const Eigen::MatrixXd& rhs, Eigen::VectorXd* q, double* resErr);
+
+  /*
+   * @return: rank of the matrix that is currently set in the linear solver
+   */
+  int getLinSolverRank() const { return linSolver_.rank(); }
```

Comment:
Can this solver return a negative rank?

Reply:
No it shouldn't. I will change it to size_t as well.

Reply:
in this context, i think `unsigned int` makes more sense, as it does not represent the size of a std container